sage_code/type_two_primes.py

```diff
@@ -53,33 +59,58 @@
 GENERIC_UPPER_BOUND = 10**30
 
 
-def LLS(p):
-    return (log(p) + 9 + 2.5 * (log(log(p))) ** 2) ** 2
+def bound_on_largest_root(d, b):
+    """From Maarten's email"""
+    return RR(exp(-d*lambert_w(-1,-1/(d*b))))
 
 
-def get_type_2_uniform_bound(ecdb_type):
+def compute_S_d(d):
 
-    if ecdb_type == "LSS":
-        BOUND_TERM = (log(x) + 9 + 2.5 * (log(log(x))) ** 2) ** 2
-    elif ecdb_type == "BS":
-        # BOUND_TERM = (4*log(x) + 10)**2
-        # BOUND_TERM = (3.29*log(x) + 2.96 + 4.9)**2
-        BOUND_TERM = (1.881 * log(x) + 2 * 0.34 + 5.5) ** 2
-    else:
-        raise ValueError("argument must be LSS or BS")
+    g = 4 * log(x) + 10
 
-    f = BOUND_TERM**6 + BOUND_TERM**3 + 1 - x
+    f_BS = x - (g ** (4*d) + g ** (2*d) + 1)
 
+    largest_root_bound = max(bound_on_largest_root(4*d, 4),bound_on_largest_root(4*d, 5))
+
+    bound_BS = find_root(f_BS, 10, largest_root_bound)
+    
+    lls = (log(x) + 9 + 2.5 * (log(log(x))) ** 2) ** 2
     try:
-        bound = find_root(f, 1000, GENERIC_UPPER_BOUND)
-        return ceil(bound)
+        bound_const = find_root(10**9 - lls, 10**6, largest_root_bound)
     except RuntimeError:
-        warning_msg = (
-            "Warning: Type 2 bound for quadratic field with "
-            "discriminant {} failed. Returning generic upper bound"
-        ).format(5)
-        print(warning_msg)
-        return GENERIC_UPPER_BOUND
+        bound_const = largest_root_bound
+
+    f_lls = x - (lls ** (4*d) + lls ** (2*d) + 1)
+    try:
+        bound_LLS = find_root(f_lls, bound_const, largest_root_bound)
+    except RuntimeError:
+        bound_LLS = largest_root_bound
+
+    return ceil(min(bound_BS, max(bound_const, bound_LLS))) 
+
+
+def improved_BS(d, A=4, B=2.5, C=5):
+    """Tool for helping do the calculation in Example 6.6"""
+    g = A * log(x) + 2*B + C
+
+    f_BS = x - (g ** (4*d) + g ** (2*d) + 1)
+
+    largest_root_bound = max(bound_on_largest_root(4*d, 4),bound_on_largest_root(4*d, 5))
+
+    bound_BS = find_root(f_BS, 10, largest_root_bound)
+
```

Comment:
again not sure where this comes from ...

Reply:
It is just a number that comes from some computation. It could probably be improved, however it doesn't seem necessary since this number is small enough in practice.

We want to proof that for x>c we have (4log(x)+2*B+C)^(4d) + (4log(x)+2*B+C)^(2d) +1 <=  (5log(x))^(4d) . We need this inequality in order to ensure that bound_on_largest_root(4*d, 5) is indeed large enough. I claim we can take c = exp(2*B+C+1). Indeed we have 
(4log(x)+2*B+C)^(4d) + (4log(x)+2*B+C)^(2d) +1 <= (4log(x)+2*B+C+1)^(4d) <= (5log(x))^(4d)

where the first inequality comes from just expanding (4log(x)+2*B+C+1)^(4d) and the second inequality is because log(x) >= 2*B+C+1 as soon as x >  exp(2*B+C+1).

Reply:
ok fair enough ...